mne_bids/path.py

```diff
@@ -120,22 +149,77 @@ def __str__(self):
                     key = 'rec'
                 basename.append('%s-%s' % (key, val))
 
-        if self.suffix is not None:
-            basename.append(self.suffix)
+        if self.kind is not None:
+            if self.extension is not None:
+                basename.append(f'{self.kind}{self.extension}')
+            else:
+                basename.append(self.kind)
 
         basename = '_'.join(basename)
-        if self.prefix is not None:
-            basename = op.join(self.prefix, basename)
-
         return basename
 
+    @property
+    def fpath(self):
+        """Full filepath for this BIDS file.
+
+        Getting the file path consists of the following behavior:
+
+        1. ``bids_root`` is None: A warning is shown
+        to the user and it returns the ``basename``.
+
+        2. ``bids_root`` is not None: The full file path
+        is inferred, or an error is raised if not possible.
+
+        Returns
+        -------
+        bids_fpath : str
+            Either the ``basename``, or full filepath to the dataset.
+            If ``kind`` and ``extension`` are not explicitly given,
+            the full file path will be inferred.
+
+            If it cannot find a matching dataset, it
+            will result in an error.
+        """
+        if self.bids_root is None:
+            msg = ('Bids root is None and '
+                   'full BIDS filepath cannot be '
+                   'automatically inferred. '
+                   'Returning the bids basename.')
+            warn(msg)
+            return self.basename
+
+        # allows functionality as a file path that doesn't
+        # exist yet (e.g. in write_raw_bids)
+        # TODO: HACK to allow file paths that don't exist yet.
+        if self.kind is not None and self.extension is not None:
+            return op.join(self.bids_root, self.basename)
```

Comment:
How come you can simply join the bids root and the basename to create a valid path, shouldn't you have to insert subject, session, kind subfolders too? Or am I missing something?

cc @agramfort

Reply:
I see, so inserting `sub-<id>/` into the path if `subject is not None` and doing the same for `session` and `kind`. That makes sense and I can add that in. Lmk if I'm interpreting correctly or not.

**For path inference**
So basically if anything besides 'subject', 'session' and 'task' are defined, then it will try to infer the full filepath based on what is available. E.g. if `task='01'` and there is only one dataset with task '01' and everything else is 02, it will find that one dataset, else, it will return an error asking for more to be defined?

Reply:
Okay just tried it and it's not trivial. There are a lot of places in the codebase and unit tests where there are edge cases where they utilize the fact that `prefix` was not necessarily the bids_root, but actually the parent directory of that intended filepath. For example in `copyfile_kit`:

```Python
    data_path = op.split(dest)[0]
    if 'mrk' in _init_kwargs:
        hpi = _init_kwargs['mrk']
        acq_map = dict()
        if isinstance(hpi, list):
            if _get_mrk_meas_date(hpi[0]) > _get_mrk_meas_date(hpi[1]):
                raise ValueError('Markers provided in incorrect order.')
            _, marker_ext = _parse_ext(hpi[0])
            acq_map = dict(zip(['pre', 'post'], hpi))
        else:
            _, marker_ext = _parse_ext(hpi)
            acq_map[None] = hpi
        for key, value in acq_map.items():
            marker_fname = make_bids_basename(
                subject=subject_id, session=session_id, task=task,
                acquisition=key, run=run,
                bids_root=data_path, suffix=f'markers{marker_ext}')
            sh.copyfile(value, marker_fname)
```

Here, by making `bids_root=data_path`, we will get problems. I could refactor all of these areas to utilize `bids_root` explicitly, but idk how much changes is too much. Any guidance?

Reply:
Refactored now adding this snippet to `fpath`:

```Python
# create the data path based on entities available
        data_path = self.bids_root
        if self.subject is not None:
            data_path = op.join(data_path, f'sub-{self.subject}')
        if self.session is not None:
            data_path = op.join(data_path, f'ses-{self.session}')

        # file-kind will allow 'meg', 'eeg', 'ieeg', 'anat'
        if self.kind is None:
            msg = ('No kind was provided, and it cannot be '
                   'automatically inferred. Please set kind to one of '
                   f'{ALLOWED_FILENAME_KINDS}.')
            raise RuntimeError(msg)
        elif self.kind in ALLOWED_MODALITY_KINDS:
            data_path = op.join(data_path, self.kind)
        elif self.kind == 'T1w':
            data_path = op.join(data_path, 'anat')
```